platform/mbed_critical.c

```diff
@@ -51,53 +51,44 @@ bool core_util_is_isr_active(void)
 #endif
 }
 
-MBED_WEAK void core_util_critical_section_enter(void)
+bool core_util_in_critical_section(void)
 {
-    bool interrupts_disabled = !core_util_are_interrupts_enabled();
-    __disable_irq();
-
-    /* Save the interrupt disabled state as it was prior to any nested critical section lock use */
-    if (!interrupt_enable_counter) {
-        critical_interrupts_disabled = interrupts_disabled;
-    }
+    return (critical_section_reentrancy_counter != 0);
+}
 
-    /* If the interrupt_enable_counter overflows or we are in a nested critical section and interrupts
-       are enabled, then something has gone badly wrong thus assert an error.
-    */
-    MBED_ASSERT(interrupt_enable_counter < UINT32_MAX); 
+void core_util_critical_section_enter(void)
+{
 // FIXME
-#ifndef   FEATURE_UVISOR
-    if (interrupt_enable_counter > 0) {
-        MBED_ASSERT(interrupts_disabled);
-    }
-#else
-#warning "core_util_critical_section_enter needs fixing to work from unprivileged code"
+#ifdef FEATURE_UVISOR
+    #warning "core_util_critical_section_enter needs fixing to work from unprivileged code"
 #endif /* FEATURE_UVISOR */
-    interrupt_enable_counter++;
+
+    // If the reentrancy counter overflows something has gone badly wrong.
+    MBED_ASSERT(critical_section_reentrancy_counter < UINT32_MAX);
+
+    if (critical_section_reentrancy_counter == 0) {
```

Comment:
please note that at the check time we have not yet entered a critical section. This can lead to more than one call to  `hal_critical_section_enter` which contains `critical_interrupts_enabled ` global variable. As a result this variable can be overwritten leading to a bug.